api-spec/filters.md

```diff
@@ -0,0 +1,133 @@
+Before is `{"time":{"lt":"2018-03-20T16:11:44.353Z"}}`  
+After is `{"time":{"gt":"2018-03-20T16:11:44.353Z"}}`  
+Before with Equality is `{"time":{"lte":"2018-03-20T16:11:44.353Z"}}`  
+After with Equality is `{"time":{"gte":"2018-03-20T16:11:44.353Z"}}`  
+
+Filter Extensions
+-----------------
+
+Spatial and Temporal filtering is not enough for many use cases. STAC is designed to be extensible, so we encourage additional filters to be provided as extensions.
+
+Filter extensions are intended to provide a place to specify additional filter types and behavior. Some filters may need to target vendor specific fields or use a more complex structure when defining the filter. For example, a geospatial intersection filter that supports a polygon as input.
+
+Filter extensions should be specified as an openapi fragment. This is a section of YAML that can be incorporated into the larger api definition. These fragments can then be combined into an implementation specific api definition that describes the full api.
+
+Example fragment:
+```
+  IntersectsFilter:
+    type: object
+    description: >-
+          Only returns items that intersect with the provided polygon.
+    properties:
+      intersects:
+        $ref: '#/definitions/Polygon'
+  Polygon:
+    type: object
+    properties:
+      type:
+        type: string
+        enum:
+          - polygon
+      coordinates:
+        $ref: '#/definitions/Polygon2D'
+    required:
+      - type
+      - coordinates
+  Polygon2D:
+    type: array
+    minItems: 1
+    items:
+      $ref: '#/definitions/LinearRing2D'
+  LinearRing2D:
+    type: array
+    minItems: 4
+    items:
+      $ref: '#/definitions/Point2D'
+    example: [[-104,35],[-103,35],[-103,34],[-104,34],[-104,35]]
+  Point2D:
+    description: A 2d geojson point
+    type: array
+    minimum: 2
+    maximum: 2
+    items:
+      type: number
+      format: double
+    example:
+      - -104
+      - 35
+```
+
+It is likely that there are schemas that should be used in common for types of filters that target different fields. We should define a common set of filter types that can be used in defining filters for different fields.
+- NumberRange
+- TimeRange
+- Text
+- ArrayIncludes
+- Etc
+
+When defining a new filter fragment you would reference these common filter types.
+```
+CloudCover:
+    type: object
+    description: >-
+          Filter items by desired cloud coverage.
+    properties:
+      cloudcover:
+        $ref: '#/definitions/NumberRange'
+```
+
+Some additional extensions that have been discussed:
+
+CQL support for generic queries:  
+`{"CQL": "CQL Select String"}`
+
+Time intervals:  
+`{"time": "P1Y"}`  Assume Duration/now if no time specified?  
+`{"time": "2018/P1M"}` Any time in january of 2018  
+
+
+Adding filters to search
+------------------------
+
+I am hoping that we can agree to support openapi3 syntax going forward. It would allow a simple addition to the searchBody portion using allOf:
+```
```

Comment:
I think this would be ok, but I do think it is important to provide some form of swagger / openapi approximation. What would that look like? Like what's the 'alternative' with swagger? I'm ok if it's leaving off a decent chunk of functionality, as long as we just explain that.

Reply:
I didn't provide a swagger 2.0 version, but it would be something like:

definitions:
  searchBody:
    description: The search criteria
    type: object
    properties:
      bbox:
        $ref: '#/definitions/bbox'
  bbox:
    type: array
    items:
      type: number